File: mobile/lib/shared/models/ownership_verification.dart
import 'property.dart';
import 'user.dart';
import 'organization.dart';
import 'booking.dart';
import 'reservation.dart';
import 'vacation_rental.dart';

enum OwnershipVerificationStatus {
  PENDING,
  VERIFIED,
  REJECTED,
  EXPIRED,
}

enum VerificationMethod {
  MANUAL,
  API,
  BLOCKCHAIN,
  AI,
}

class PropertyOwnershipVerification {
  final String id;
  final String propertyId;
  final String orgId;
  final String? currentOwnerId;
  final OwnershipVerificationStatus verificationStatus;
  final VerificationMethod verificationMethod;
  final DateTime? verifiedAt;
  final String? verifiedBy;
  final DateTime? expiresAt;
  final String? rejectionReason;
  final String? governmentTransactionId;
  final double? aiConfidenceScore;
  final bool manualReviewRequired;
  final bool priorityVerification;
  final String? verificationNotes;
  final Map<String, dynamic>? supportingDocuments;
  final Map<String, dynamic>? ownershipHistory;
  final String? legalDescription;
  final String? parcelNumber;
  final String? jurisdiction;
  final DateTime? recordingDate;
  final Map<String, dynamic>? chainOfCustody;
  final Map<String, dynamic>? verificationMetadata;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  // Relations
  final Property? property;
  final Organization? organization;
  final User? currentOwner;
  final User? verifier;
  final List<OwnershipVerificationDocument>? documents;
  final List<Booking>? bookings;
  final List<Reservation>? reservations;
  final List<VacationRental>? vacationRentals;

  const PropertyOwnershipVerification({
    required this.id,
    required this.propertyId,
    required this.orgId,
    this.currentOwnerId,
    required this.verificationStatus,
    required this.verificationMethod,
    this.verifiedAt,
    this.verifiedBy,
    this.expiresAt,
    this.rejectionReason,
    this.governmentTransactionId,
    this.aiConfidenceScore,
    required this.manualReviewRequired,
    required this.priorityVerification,
    this.verificationNotes,
    this.supportingDocuments,
    this.ownershipHistory,
    this.legalDescription,
    this.parcelNumber,
    this.jurisdiction,
    this.recordingDate,
    this.chainOfCustody,
    this.verificationMetadata,
    required this.createdAt,
    required this.updatedAt,
    this.property,
    this.organization,
    this.currentOwner,
    this.verifier,
    this.documents,
    this.bookings,
    this.reservations,
    this.vacationRentals,
  });

  factory PropertyOwnershipVerification.fromJson(Map<String, dynamic> json) {
    return PropertyOwnershipVerification(
      id: json['id'] as String,
      propertyId: json['propertyId'] as String,
      orgId: json['orgId'] as String,
      currentOwnerId: json['currentOwnerId'] as String?,
      verificationStatus: OwnershipVerificationStatus.values.firstWhere(
        (e) => e.name == json['verificationStatus'],
        orElse: () => OwnershipVerificationStatus.PENDING,
      ),
      verificationMethod: VerificationMethod.values.firstWhere(
        (e) => e.name == json['verificationMethod'],
        orElse: () => VerificationMethod.MANUAL,
      ),
      verifiedAt: json['verifiedAt'] != null 
          ? DateTime.parse(json['verifiedAt'] as String) 
          : null,
      verifiedBy: json['verifiedBy'] as String?,
      expiresAt: json['expiresAt'] != null 
          ? DateTime.parse(json['expiresAt'] as String) 
          : null,
      rejectionReason: json['rejectionReason'] as String?,
      governmentTransactionId: json['governmentTransactionId'] as String?,
      aiConfidenceScore: (json['aiConfidenceScore'] as num?)?.toDouble(),
      manualReviewRequired: json['manualReviewRequired'] as bool,
      priorityVerification: json['priorityVerification'] as bool,
      verificationNotes: json['verificationNotes'] as String?,
      supportingDocuments: json['supportingDocuments'] as Map<String, dynamic>?,
      ownershipHistory: json['ownershipHistory'] as Map<String, dynamic>?,
      legalDescription: json['legalDescription'] as String?,
      parcelNumber: json['parcelNumber'] as String?,
      jurisdiction: json['jurisdiction'] as String?,
      recordingDate: json['recordingDate'] != null 
          ? DateTime.parse(json['recordingDate'] as String) 
          : null,
      chainOfCustody: json['chainOfCustody'] as Map<String, dynamic>?,
      verificationMetadata: json['verificationMetadata'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      property: json['property'] != null 
          ? Property.fromJson(json['property'] as Map<String, dynamic>) 
          : null,
      organization: json['organization'] != null 
          ? Organization.fromJson(json['organization'] as Map<String, dynamic>) 
          : null,
      currentOwner: json['currentOwner'] != null 
          ? User.fromJson(json['currentOwner'] as Map<String, dynamic>) 
          : null,
      verifier: json['verifier'] != null 
          ? User.fromJson(json['verifier'] as Map<String, dynamic>) 
          : null,
      documents: (json['documents'] as List<dynamic>?)
          ?.map((e) => OwnershipVerificationDocument.fromJson(e as Map<String, dynamic>))
          .toList(),
      bookings: (json['bookings'] as List<dynamic>?)
          ?.map((e) => Booking.fromJson(e as Map<String, dynamic>))
          .toList(),
      reservations: (json['reservations'] as List<dynamic>?)
          ?.map((e) => Reservation.fromJson(e as Map<String, dynamic>))
          .toList(),
      vacationRentals: (json['vacationRentals'] as List<dynamic>?)
          ?.map((e) => VacationRental.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'propertyId': propertyId,
      'orgId': orgId,
      'currentOwnerId': currentOwnerId,
      'verificationStatus': verificationStatus.name,
      'verificationMethod': verificationMethod.name,
      'verifiedAt': verifiedAt?.toIso8601String(),
      'verifiedBy': verifiedBy,
      'expiresAt': expiresAt?.toIso8601String(),
      'rejectionReason': rejectionReason,
      'governmentTransactionId': governmentTransactionId,
      'aiConfidenceScore': aiConfidenceScore,
      'manualReviewRequired': manualReviewRequired,
      'priorityVerification': priorityVerification,
      'verificationNotes': verificationNotes,
      'supportingDocuments': supportingDocuments,
      'ownershipHistory': ownershipHistory,
      'legalDescription': legalDescription,
      'parcelNumber': parcelNumber,
      'jurisdiction': jurisdiction,
      'recordingDate': recordingDate?.toIso8601String(),
      'chainOfCustody': chainOfCustody,
      'verificationMetadata': verificationMetadata,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'property': property?.toJson(),
      'organization': organization?.toJson(),
      'currentOwner': currentOwner?.toJson(),
      'verifier': verifier?.toJson(),
      'documents': documents?.map((e) => e.toJson()).toList(),
      'bookings': bookings?.map((e) => e.toJson()).toList(),
      'reservations': reservations?.map((e) => e.toJson()).toList(),
      'vacationRentals': vacationRentals?.map((e) => e.toJson()).toList(),
    };
  }

  PropertyOwnershipVerification copyWith({
    String? id,
    String? propertyId,
    String? orgId,
    String? currentOwnerId,
    OwnershipVerificationStatus? verificationStatus,
    VerificationMethod? verificationMethod,
    DateTime? verifiedAt,
    String? verifiedBy,
    DateTime? expiresAt,
    String? rejectionReason,
    String? governmentTransactionId,
    double? aiConfidenceScore,
    bool? manualReviewRequired,
    bool? priorityVerification,
    String? verificationNotes,
    Map<String, dynamic>? supportingDocuments,
    Map<String, dynamic>? ownershipHistory,
    String? legalDescription,
    String? parcelNumber,
    String? jurisdiction,
    DateTime? recordingDate,
    Map<String, dynamic>? chainOfCustody,
    Map<String, dynamic>? verificationMetadata,
    DateTime? createdAt,
    DateTime? updatedAt,
    Property? property,
    Organization? organization,
    User? currentOwner,
    User? verifier,
    List<OwnershipVerificationDocument>? documents,
    List<Booking>? bookings,
    List<Reservation>? reservations,
    List<VacationRental>? vacationRentals,
  }) {
    return PropertyOwnershipVerification(
      id: id ?? this.id,
      propertyId: propertyId ?? this.propertyId,
      orgId: orgId ?? this.orgId,
      currentOwnerId: currentOwnerId ?? this.currentOwnerId,
      verificationStatus: verificationStatus ?? this.verificationStatus,
      verificationMethod: verificationMethod ?? this.verificationMethod,
      verifiedAt: verifiedAt ?? this.verifiedAt,
      verifiedBy: verifiedBy ?? this.verifiedBy,
      expiresAt: expiresAt ?? this.expiresAt,
      rejectionReason: rejectionReason ?? this.rejectionReason,
      governmentTransactionId: governmentTransactionId ?? this.governmentTransactionId,
      aiConfidenceScore: aiConfidenceScore ?? this.aiConfidenceScore,
      manualReviewRequired: manualReviewRequired ?? this.manualReviewRequired,
      priorityVerification: priorityVerification ?? this.priorityVerification,
      verificationNotes: verificationNotes ?? this.verificationNotes,
      supportingDocuments: supportingDocuments ?? this.supportingDocuments,
      ownershipHistory: ownershipHistory ?? this.ownershipHistory,
      legalDescription: legalDescription ?? this.legalDescription,
      parcelNumber: parcelNumber ?? this.parcelNumber,
      jurisdiction: jurisdiction ?? this.jurisdiction,
      recordingDate: recordingDate ?? this.recordingDate,
      chainOfCustody: chainOfCustody ?? this.chainOfCustody,
      verificationMetadata: verificationMetadata ?? this.verificationMetadata,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      property: property ?? this.property,
      organization: organization ?? this.organization,
      currentOwner: currentOwner ?? this.currentOwner,
      verifier: verifier ?? this.verifier,
      documents: documents ?? this.documents,
      bookings: bookings ?? this.bookings,
      reservations: reservations ?? this.reservations,
      vacationRentals: vacationRentals ?? this.vacationRentals,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is PropertyOwnershipVerification &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}

class OwnershipVerificationDocument {
  final String id;
  final String verificationId;
  final String type;
  final String url;
  final String status;
  final DateTime uploadedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const OwnershipVerificationDocument({
    required this.id,
    required this.verificationId,
    required this.type,
    required this.url,
    required this.status,
    required this.uploadedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory OwnershipVerificationDocument.fromJson(Map<String, dynamic> json) {
    return OwnershipVerificationDocument(
      id: json['id'] as String,
      verificationId: json['verificationId'] as String,
      type: json['type'] as String,
      url: json['url'] as String,
      status: json['status'] as String,
      uploadedAt: DateTime.parse(json['uploadedAt'] as String),
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'verificationId': verificationId,
      'type': type,
      'url': url,
      'status': status,
      'uploadedAt': uploadedAt.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is OwnershipVerificationDocument &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}

File: client-seo/src/lib/api/ownership-verification.ts
import apiClient from "./client";



// Types
export interface OwnershipVerification {
  id: string;
  propertyId: string;
  orgId: string;
  currentOwnerId?: string;
  verificationStatus: 'PENDING' | 'VERIFIED' | 'REJECTED' | 'EXPIRED' | 'SUSPENDED';
  verificationMethod: 'DOCUMENT_UPLOAD' | 'BLOCKCHAIN_VERIFICATION' | 'GOVERNMENT_API' | 'THIRD_PARTY_SERVICE' | 'MANUAL_REVIEW' | 'AI_VERIFICATION';
  verifiedAt?: string;
  verifiedBy?: string;
  expiresAt?: string;
  rejectionReason?: string;
  governmentTransactionId?: string;
  aiConfidenceScore?: number;
  manualReviewRequired: boolean;
  priorityVerification: boolean;
  verificationNotes?: string;
  supportingDocuments?: any;
  ownershipHistory?: any;
  legalDescription?: string;
  parcelNumber?: string;
  jurisdiction?: string;
  createdAt: string;
  updatedAt: string;
  
  // Relations
  property?: {
    id: string;
    name: string;
    addressLine1: string;
    city: string;
    state: string;
    zip: string;
  };
  currentOwner?: {
    id: string;
    name: string;
    email: string;
  };
  verifier?: {
    id: string;
    name: string;
    email: string;
  };
  documents?: OwnershipVerificationDocument[];
  organization?: {
    id: string;
    name: string;
    type: string;
  };
}

export interface OwnershipVerificationDocument {
  id: string;
  verificationId: string;
  documentType: 'DEED' | 'TITLE_DEED' | 'PROPERTY_TAX_RECORD' | 'MORTGAGE_STATEMENT' | 'INSURANCE_POLICY' | 'UTILITY_BILL' | 'HOA_DOCUMENT' | 'COURT_ORDER' | 'INHERITANCE_DOCUMENT' | 'TRUST_DOCUMENT' | 'CORPORATE_RESOLUTION' | 'POWER_OF_ATTORNEY' | 'OTHER';
  fileName: string;
  filePath: string;
  fileSize: number;
  mimeType: string;
  checksum: string;
  uploadMethod: 'direct' | 'api' | 'import';
  extractedText?: string;
  extractedMetadata?: any;
  validationStatus: 'valid' | 'invalid' | 'pending';
  validationErrors?: any;
  aiAnalysisResults?: any;
  reviewedAt?: string;
  reviewedBy?: string;
  reviewNotes?: string;
  isPublic: boolean;
  accessLevel: 'private' | 'organization' | 'public';
  retentionUntil?: string;
  deletedAt?: string;
  createdAt: string;
  updatedAt: string;
  
  // Relations
  verification?: {
    id: string;
    verificationStatus: string;
    property: {
      id: string;
      name: string;
      addressLine1: string;
    };
  };
  reviewer?: {
    id: string;
    name: string;
    email: string;
  };
}

export interface PropertyOwnershipTransfer {
  id: string;
  propertyId: string;
  orgId: string;
  fromOwnerId?: string;
  toOwnerId?: string;
  transferType: 'sale' | 'inheritance' | 'gift' | 'foreclosure';
  transferStatus: 'pending' | 'completed' | 'cancelled';
  transferDate?: string;
  recordingDate?: string;
  considerationAmount?: number;
  transferDocuments?: any;
  legalDescription?: string;
  blockchainTransactionId?: string;
  governmentReference?: string;
  transferNotes?: string;
  createdAt: string;
  updatedAt: string;
  
  // Relations
  property?: {
    id: string;
    name: string;
    addressLine1: string;
  };
  organization?: {
    id: string;
    name: string;
  };
  fromOwner?: {
    id: string;
    name: string;
    email: string;
  };
  toOwner?: {
    id: string;
    name: string;
    email: string;
  };
}

export interface OwnershipVerificationAnalytics {
  summary: {
    total: number;
    verified: number;
    pending: number;
    rejected: number;
    expired: number;
    verificationRate: string;
    averageConfidence: string;
  };
  byMethod: Array<{
    verificationMethod: string;
    _count: number;
  }>;
  recent: OwnershipVerification[];
}

// API Client
export const ownershipVerificationApi = {
  // Get all ownership verifications
  getVerifications: async (params?: {
    page?: number;
    limit?: number;
    propertyId?: string;
    orgId?: string;
    verificationStatus?: string;
    verificationMethod?: string;
  }): Promise<{
    verifications: OwnershipVerification[];
    pagination: {
      page: number;
      limit: number;
      total: number;
      totalPages: number;
    };
  }> => {
    const queryParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          queryParams.append(key, value.toString());
        }
      });
    }
    
    return await apiClient.get(`/property-ownership-verification?${queryParams.toString()}`);
  },

  // Get single ownership verification
  getVerificationById: async (id: string): Promise<{ data: OwnershipVerification }> => {
    return await apiClient.get(`/property-ownership-verification/${id}`);
  },

  // Create ownership verification
  createVerification: async (data: {
    propertyId: string;
    orgId: string;
    currentOwnerId?: string;
    verificationMethod: 'DOCUMENT_UPLOAD' | 'BLOCKCHAIN_VERIFICATION' | 'GOVERNMENT_API' | 'THIRD_PARTY_SERVICE' | 'MANUAL_REVIEW' | 'AI_VERIFICATION';
    verificationNotes?: string;
    supportingDocuments?: any;
    legalDescription?: string;
    parcelNumber?: string;
    jurisdiction?: string;
  }): Promise<{ data: OwnershipVerification }> => {
    return await apiClient.post('/property-ownership-verification', data);
  },

  // Update verification status
  updateVerificationStatus: async (id: string, data: {
    verificationStatus: 'PENDING' | 'VERIFIED' | 'REJECTED' | 'EXPIRED' | 'SUSPENDED';
    rejectionReason?: string;
    aiConfidenceScore?: number;
  }): Promise<{ data: OwnershipVerification }> => {
    return await apiClient.put(`/property-ownership-verification/${id}/status`, data);
  },

  // Upload verification document
  uploadDocument: async (verificationId: string, data: {
    documentType: 'DEED' | 'TITLE_DEED' | 'PROPERTY_TAX_RECORD' | 'MORTGAGE_STATEMENT' | 'INSURANCE_POLICY' | 'UTILITY_BILL' | 'HOA_DOCUMENT' | 'COURT_ORDER' | 'INHERITANCE_DOCUMENT' | 'TRUST_DOCUMENT' | 'CORPORATE_RESOLUTION' | 'POWER_OF_ATTORNEY' | 'OTHER';
    fileName: string;
    filePath: string;
    fileSize: number;
    mimeType: string;
    checksum: string;
    uploadMethod: 'direct' | 'api' | 'import';
    extractedText?: string;
    extractedMetadata?: any;
    accessLevel?: 'private' | 'organization' | 'public';
    retentionUntil?: string;
  }): Promise<{ data: OwnershipVerificationDocument }> => {
    return await apiClient.post(`/property-ownership-verification/${verificationId}/documents`, data);
  },

  // Review document
  reviewDocument: async (documentId: string, data: {
    validationStatus: 'valid' | 'invalid' | 'pending';
    validationErrors?: any;
    reviewNotes?: string;
  }): Promise<{ data: OwnershipVerificationDocument }> => {
    return await apiClient.put(`/property-ownership-verification/documents/${documentId}/review`, data);
  },

  // Get verification transfers
  getTransfers: async (propertyId: string): Promise<{ data: PropertyOwnershipTransfer[] }> => {
    return await apiClient.get(`/property-ownership-verification/${propertyId}/transfers`);
  },

  // Create ownership transfer
  createTransfer: async (propertyId: string, data: {
    orgId: string;
    fromOwnerId?: string;
    toOwnerId?: string;
    transferType: 'sale' | 'inheritance' | 'gift' | 'foreclosure';
    transferDate?: string;
    considerationAmount?: number;
    transferDocuments?: any;
    legalDescription?: string;
    blockchainTransactionId?: string;
    governmentReference?: string;
    transferNotes?: string;
  }): Promise<{ data: PropertyOwnershipTransfer }> => {
    return await apiClient.post(`/property-ownership-verification/${propertyId}/transfers`, data);
  },

  // Get verification analytics
  getAnalytics: async (params?: {
    orgId?: string;
    startDate?: string;
    endDate?: string;
  }): Promise<OwnershipVerificationAnalytics> => {
    const queryParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          queryParams.append(key, value.toString());
        }
      });
    }
    
    return await apiClient.get(`/property-ownership-verification/analytics/summary?${queryParams.toString()}`);
  },

  // Helper method to upload file with checksum
  uploadFileWithChecksum: async (
    verificationId: string,
    file: File,
    documentType: OwnershipVerificationDocument['documentType'],
    options?: {
      extractedText?: string;
      extractedMetadata?: any;
      accessLevel?: 'private' | 'organization' | 'public';
      retentionUntil?: string;
    }
  ): Promise<{ data: OwnershipVerificationDocument }> => {
    // Generate checksum
    const buffer = await file.arrayBuffer();
    const checksum = await crypto.subtle.digest('SHA-256', buffer);
    const checksumHex = Array.from(new Uint8Array(checksum))
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');

    // Upload file (this would typically go through a file upload endpoint first)
    // For now, we'll assume the file is uploaded and we have the path
    const filePath = `/uploads/ownership-documents/${file.name}`;
    
    return await ownershipVerificationApi.uploadDocument(verificationId, {
      documentType,
      fileName: file.name,
      filePath,
      fileSize: file.size,
      mimeType: file.type,
      checksum: checksumHex,
      uploadMethod: 'direct',
      extractedText: options?.extractedText,
      extractedMetadata: options?.extractedMetadata,
      accessLevel: options?.accessLevel,
      retentionUntil: options?.retentionUntil
    });
  }
};

export default ownershipVerificationApi;
